src/dune_pkg/fetch.mli

```diff
@@ -1,20 +1,23 @@
 open Import
 
 type failure =
-  | Checksum_mismatch of Checksum.t
+  | Checksum_mismatch of
+      { expected : Checksum.t
+      ; got : Checksum.t
+      }
   | Unavailable of User_message.t option
 
-(** [fetch ~checksum ~target url] will fetch [url] into [target]. It will verify
-    the downloaded file against [checksum], unless it [checksum] is [None].
+(** [fetch ~checksums ~target url] will fetch [url] into [target]. It will verify
+    the downloaded file against [checksums], unless it [checksums] is [None].
 
     return [Error (Checksum_mismatch _)] When the downloaded file doesn't match
-    the expected [checksum], this will pass the actually computed checksum.
+    the expected [checksums], this will return the actually computed checksum.
 
     return [Error (Unavailable _))] When the file can't be retrieved, e.g. not
     available at the location. *)
 val fetch
   :  unpack:bool
-  -> checksum:Checksum.t option
+  -> checksums:Checksum.t Nonempty_list.t option
```

Comment:
I think this is equivalent to just a list of checksums

Reply:
I know, but what I want to express is that either there are checksums in which case there should be at least one or there is specifically no checksum. This avoids the case where "there are checksums" but the checksums in question are an empty list.

Reply:
I don't understand the difference between

> specifically no checksum

and

>  "there are checksums" but the checksums in question are an empty list.

And it's not for the lack of trying. I suggest that I as a reader don't see the difference, it's not going to be obvious to many other reads either. In which case, no additional clarity is gained here.

Reply:
The distinction between "no content" and "there is at least one thing" is in my eyes a pretty core part of `option` types. Because we can of course always replace `None` with `[]` and `Some x` with `[x]` but the option type signifies that the absence of content triggers a different behavior than its presence. In the dune codebase I have the impression that we try to be explicit that something is missing, e.g. through `Loc.none` in the case of location info.

If we just have a list, it is easy to assume you have checksums but the list of checksums is 0 thus you don't actually have checksums. Empty lists operate like lists with content under most list operators so it is easy to pass an empty list where you didn't mean to because the previous function that computed you the checksums served you with an empty list (e.g. incorrectly as result of a bug).

This is also why I use `Nonempty_list.t` because otherwise it would be hard to understand whether there is a semantic difference between `None` and `Some []`.

Reply:
I removed it now, but we now have `fetch ~checksums:[]` and `fetch_without_checksum` which seem to be doing roughly the same. Maybe the right way would be to preserve the `Nonempty_list.t` and use `fetch_without_checksum` in places where we don't expect checksums.

Reply:
I think I understand where you were going with this, but I think that clarifying the distinction can only be done with an explicit constructor name. E.g. `No_checksums | At_least of Nonempty_list.t`. Otherwise, you just replaced Cons and Nil with Some, and None.

FTR, this is my interpretation of how the function works:

It takes a list of checksums and verifies all of them.
If there are no checksums in the list, there are no checksums to verify.
If there's one checksum in the list, there is 1 checksum to verify.
If there's two checksums in the list, ...

Reply:
Fair enough point about the constructor, I agree that that would make it more explicit than `Some (Cons ...)`. In my mind one of the reasons we had a distinction between `fetch` and `fetch_without_checksums` is to make sure that if the user picked `fetch` would have (at least one) checksum(s) passed.